README.md
# Map's View

Map's View is a versatile Python application designed to help users explore geographical data interactively. With its intuitive user interface and comprehensive feature set, Map's View is an ideal tool for visualizing and analyzing spatial information.

---

![](documentation_images/Example_1.png)

`documentation_images/Example_1.png`

## Features

- **Multi-layered Map Display:** Map's View provides a dynamic map display with support for multiple layers. You can overlay different map types and toggle between them to gain different perspectives on your data.

- **Zoom and Pan Controls:** Effortlessly navigate through the map using standard controls for zooming in and out and panning across different regions. This allows for detailed exploration of specific areas of interest.

- **Search Functionality:** Easily search for places by name, address, or coordinates. Map's View leverages powerful search algorithms to quickly locate specific locations on the map.

- **Marker Placement:** Add markers to the map to mark important points or points of interest. This feature is particularly useful for annotating specific locations and making them stand out on the map.

- **Layer Styles:** Customize the appearance of map layers with different styles and color schemes. Tailor the visualization to suit your preferences or specific requirements.

- **Offline Maps:** Map's View supports offline maps, allowing you to work with geographical data even when you don't have an internet connection. This feature ensures that you can access and analyze your data regardless of your location.

---

![](documentation_images/Example_2.gif)

`documentation_images/Example_2.gif`

## Installation

To install Map's View, follow these steps:

1. Clone the repository: `git clone https://github.com/EnthusiasticXcoder/Maps-View.git`
2. Navigate to the project directory: `cd Maps-View`
3. Install the required dependencies: `pip install -r requirements.txt`

## Usage

To run Map's View, execute the following command:

```shell
python main.py
```

Upon launching the application, you will be greeted with the interactive map interface. Use the provided controls and features to explore and analyze your geographical data.

## Examples

Here are a few code snippets that demonstrate the usage of specific features in Map's View:

#### 1. Searching in the map:

https://github.com/EnthusiasticXcoder/Attendence-Management-Software/assets/115250815/d048f1ae-600a-43c4-82cc-d277eb95aa4c

---

#### 2. Changing the layer style:

https://github.com/EnthusiasticXcoder/Maps-View/assets/115250815/e8510fbe-f655-43cf-99f2-1c74699e01b5

---

#### 3. Adding a marker to the map:

https://github.com/EnthusiasticXcoder/Maps-View/assets/115250815/ef84325b-1db7-488f-b289-1d15c9d9ba31


Feel free to explore the codebase and experiment with different functionalities to make the most out of Map's View.

## Contributing

Contributions are welcome! If you have any ideas, suggestions, or bug reports, please create an issue or submit a pull request.

## License

Map's View is licensed under the [MIT License](LICENSE).

History.txt
indore
paris
london
india
nyc
ujjain


App/Map.py
import tkinter
import tkintermapview as CTkMap 


class Mapwidget(CTkMap.TkinterMapView):
    def __init__(self, *args, width: int = 300, height: int = 200, 
                 corner_radius: int = 0, bg_color: str = None, 
                 database_path: str = None, use_database_only: bool = False, 
                 max_zoom: int = 19, **kwargs):
        
        super().__init__(*args, width=width, height=height, 
                         corner_radius=corner_radius, 
                         bg_color=bg_color, database_path=database_path, 
                         use_database_only=use_database_only, 
                         max_zoom=max_zoom, **kwargs)

        self.marker_list = []
        self.search_history = []
        # load markers from file
        self.load_data()

        # Set default values
        self.set_address("Indore")
        self.add_left_click_map_command(self.click_zoom)
    
    def load_data(self):
        with open('History.txt','r') as file:
            data = file.readlines()
            for history in data :
                self.search_history.append(history.replace("\n",""))

        with open('data.txt','r') as file:
            marker_data=file.readlines()
            for data in marker_data :
                data=data.replace('\n',"")
                data=data.split(',')
                if data[3]=='None' and data[4]=='None':
                    self.marker_list.append(self.set_marker(deg_x=float(data[0]),deg_y=float(data[1]),text=data[2], command=self.click_marker_event))
                elif data[3]!='None':
                    self.marker_list.append(self.set_marker(deg_x=float(data[0]),deg_y=float(data[1]), text=data[2], 
                                image=data[3],image_zoom_visibility=(0, float("inf")), command=self.click_marker_event))
                elif data[4]!='None':
                    self.marker_list.append(self.set_marker(deg_x=float(data[0]),deg_y=float(data[1]), text=data[2], 
                                image=data[3],icon=data[4],image_zoom_visibility=(0, float("inf")), command=self.click_marker_event))
        
    def search_event(self, address : str | None, event=None):
        if not self.search_history.count(address) :
            self.search_history.append(address)
        self.set_address(address)
    
    def Clear_Search_History(self):
        with open('History.txt','w') as file :
            file.write("")
        self.search_history = []
    
    def Remove_search(self,Element: str = None):
        self.search_history.remove(Element)

    def click_marker_event(self,marker):
        def Hide_Show_Image():
            if marker.image_hidden is True:
                marker.hide_image(False)
            else:
                marker.hide_image(True)

        def MarkerText():
            pass

        def MarkerImage():
            pass

        m = tkinter.Menu(self, tearoff=0)
        m.add_command(label='Edit Text', command=lambda:MarkerText)
        m.add_command(label="Edit Image", command=lambda:MarkerImage)
        m.add_command(label="Hide/Show Image", command=lambda:Hide_Show_Image)

        position=marker.get_canvas_pos(marker.position)

        m.tk_popup(int(position[0]+self.winfo_rootx()),int(position[1]+self.winfo_rooty()))

        
    def add_marker(self,coords=None,marker_text=None):
        current_position = self.get_position() if coords == None else coords
        adr = CTkMap.convert_coordinates_to_address(current_position[0], current_position[1]) 
        marker_name= marker_text if marker_text!="" or " " else f'{adr.city}, {adr.state}, {adr.country}'
        self.marker_list.append(self.set_marker(current_position[0], current_position[1], text=marker_name, command=self.click_marker_event))
    
    def Remove_marker(self, Object_Marker):
        self.marker_list.remove(Object_Marker)

    def clear_markers(self):
        for marker in self.marker_list:
            marker.delete()
        self.marker_list=[]
        with open('data.txt','w') as file:
            file.write("")

    def click_zoom(self,coordinate):
        self.set_position(coordinate[0],coordinate[1])
        self.set_zoom(zoom=self.last_zoom+1)
    
    def change_map(self, new_map: str):
        if new_map == "OpenStreetMap":
            self.set_tile_server("https://a.tile.openstreetmap.org/{z}/{x}/{y}.png")
        elif new_map == "Google normal":
            self.set_tile_server("https://mt0.google.com/vt/lyrs=m&hl=en&x={x}&y={y}&z={z}&s=Ga", max_zoom=22)
        elif new_map == "Google satellite":
            self.set_tile_server("https://mt0.google.com/vt/lyrs=s&hl=en&x={x}&y={y}&z={z}&s=Ga", max_zoom=22)
        elif new_map == "Paint Style":   
            self.set_tile_server("http://c.tile.stamen.com/watercolor/{z}/{x}/{y}.png")
        elif new_map == "Black & White":
            self.set_tile_server("http://a.tile.stamen.com/toner/{z}/{x}/{y}.png")

if __name__=='__main__':
    import customtkinter as ctk
    r=ctk.CTk()
    Mapwidget(r).pack(fill='both',expand=1)
    r.mainloop()

App/main.py
import customtkinter as ctk
import mapapp


app = mapapp.App()
def marker_box(coords=None):
    markerInput=ctk.CTkInputDialog(title='Marker Name',text="Marker Input")
    if coords != None :
        app.MapView.add_marker(coords=coords,marker_text=markerInput.get_input())
    else : 
        app.MapView.add_marker(marker_text=markerInput.get_input())
    app.Configure_marker_frame()
    
def Clear_Marker():
    app.MapView.clear_markers()
    app.Configure_marker_frame()
    
def Search_Location(event = None):
    app.MapView.search_event(app.SearchFrame.Search_Entry.get())
    app.SearchFrame.Search_Entry.delete(0,'end')
    app.Search_Frame()

def clear_Command():
    app.MapView.Clear_Search_History()
    app.Search_Frame()
    
def Marker_Location(event = None):
    app.MapView.search_event(app.Search_Entry.get())
    app.Search_Entry.delete(0,'end')

app.MapView.add_right_click_menu_command("Set Marker",command=marker_box,pass_coords=True)
    
app.SearchFrame.Search_Entry.bind("<Return>", Search_Location)
app.SearchFrame.Search_Button.configure(command=Search_Location)

app.SetMarkerButton.configure(command=marker_box)
app.RemoveMarkerButton.configure(command=Clear_Marker)
app.Search_Entry.bind("<Return>", Marker_Location )

app.SearchFrame.ClearButton.configure(command = clear_Command)

app.start()
with open('data.txt','w') as file:
    for marker in app.MapView.marker_list:
        file.write(f'{marker.position[0]},{marker.position[1]},{marker.text},{marker.image},{marker.icon}\n')
with open('History.txt','w') as file :
    for history in app.MapView.search_history :
        file.write(f"{history}\n")

App/HelperFrame.py
import customtkinter as ctk
from PIL import Image
from typing import Tuple


class Search_Frame(ctk.CTkFrame):
    def __init__(self, master: any,
                 width: int = 200, height: int = 200,
                 corner_radius: int | str | None = None, 
                 border_width: int | str | None = None, 
                 bg_color: str | Tuple[str, str] = "transparent", 
                 fg_color: str | Tuple[str, str] | None = None, 
                 border_color: str | Tuple[str, str] | None = None,
                 background_corner_colors: Tuple[str | Tuple[str, str]] | None = None, 
                 overwrite_preferred_drawing_method: str | None = None, **kwargs):
       
        super().__init__(master, 
                        width = width, height = height, 
                        corner_radius = corner_radius, 
                        border_width = border_width, 
                        bg_color = bg_color, 
                        fg_color = fg_color, 
                        border_color = border_color, 
                        background_corner_colors = background_corner_colors, 
                        overwrite_preferred_drawing_method = None, **kwargs)
        
        self.grid(row=1,column=0,sticky='nsew')

        self.Search_Entry = ctk.CTkEntry(master=self,width=250,height=35,
                                            placeholder_text="Type Address")
        self.Search_Entry.grid(row=0, column=0, sticky="we", padx=(12, 0), pady=12)

        self.Search_Button = ctk.CTkButton(master=self,text="Search",width=90)
        self.Search_Button.grid(row=0, column=1, sticky="w", padx=(12,10), pady=12)
        
        # Clear history frame
        clearFrame = ctk.CTkFrame(master=self,fg_color='transparent')
        clearFrame.grid(row=1,column=0,sticky='nsew',columnspan=2)

        clearFrame.grid_columnconfigure(0,weight=1)
        clearFrame.grid_columnconfigure(1,weight=0)
        
        ctk.CTkFrame(master=clearFrame,fg_color='black',height=2).grid(row=0,column=0,padx=(5,1),sticky='ew')

        self.ClearButton = ctk.CTkButton(master=clearFrame,cursor='hand2',
                        text="Clear all",hover=False,width=0,
                        fg_color='transparent',text_color=('#01719D',"skyblue"),border_width=0,
                        font=("times new roman",18))
        self.ClearButton.grid(row=0,column=1,padx=(2,8))


class Layer_Frame(ctk.CTkFrame):
    def __init__(self, master: any,text : any,font : ctk.CTkFont | None,command : any,
                 width: int = 200, height: int = 200,
                 corner_radius: int | str | None = None, 
                 border_width: int | str | None = None, 
                 bg_color: str | Tuple[str, str] = "transparent", 
                 fg_color: str | Tuple[str, str] | None = 'transparent', 
                 border_color: str | Tuple[str, str] | None = None,
                 background_corner_colors: Tuple[str | Tuple[str, str]] | None = None, 
                 overwrite_preferred_drawing_method: str | None = None, **kwargs):
       
        super().__init__(master, 
                        width = width, height = height, 
                        corner_radius = corner_radius, 
                        border_width = border_width, 
                        bg_color = bg_color, 
                        fg_color = fg_color, 
                        border_color = border_color, 
                        background_corner_colors = background_corner_colors, 
                        overwrite_preferred_drawing_method = None, **kwargs)
        
        self.columnconfigure(0,weight=1)
        button =ctk.CTkButton(master=self,text=text,font=font, text_color=('grey5','grey80'),
                              width=30,compound='left',fg_color='transparent',anchor='sw',hover_color=('grey60','grey25'),
                              command=lambda : command(text))
        button.grid(row=0,column=0,sticky='ew')
        
class Marker(ctk.CTkFrame):
    def __init__(self, master: any,marker : any,font : ctk.CTkFont | None,command : any,
                 width: int = 200, height: int = 200,
                 corner_radius: int | str | None = None, 
                 border_width: int | str | None = None, 
                 bg_color: str | Tuple[str, str] = "transparent", 
                 fg_color: str | Tuple[str, str] | None = 'transparent', 
                 border_color: str | Tuple[str, str] | None = None,
                 background_corner_colors: Tuple[str | Tuple[str, str]] | None = None, 
                 overwrite_preferred_drawing_method: str | None = None, **kwargs):
       
        super().__init__(master, 
                        width = width, height = height, 
                        corner_radius = corner_radius, 
                        border_width = border_width, 
                        bg_color = bg_color, 
                        fg_color = fg_color, 
                        border_color = border_color, 
                        background_corner_colors = background_corner_colors, 
                        overwrite_preferred_drawing_method = None, **kwargs)
        self.columnconfigure(0,weight=1)
        self.marker_refrence=ctk.CTkButton(self,text=marker.text,font=font, text_color=('grey10','grey80'),
                    width=30,fg_color='transparent',anchor='sw',hover_color=('grey60','grey25'),
                    command=lambda: command[0](marker.position[0],marker.position[1]))
        self.marker_refrence.grid(row=0,column=0,sticky='ew')
        
        self.XButton = ctk.CTkButton(master=self,text="",fg_color='transparent',hover_color='red',width=20,
                                     command= lambda : self.XCommand(command , marker),
                                    image=ctk.CTkImage(dark_image=Image.open("App/Images/light mode/x.png"),
                                                       light_image=Image.open("App/Images/dark mode/x.png")))
        self.XButton.grid(row=0,column=1,sticky='e')
    
    def XCommand(self,command, Element):
        command[1](Element)
        command[2]()

class History(ctk.CTkFrame):
    def __init__(self, master: any,text : any,font : ctk.CTkFont | None,command : any,
                 width: int = 200, height: int = 200,
                 corner_radius: int | str | None = None, 
                 border_width: int | str | None = None, 
                 bg_color: str | Tuple[str, str] = "transparent", 
                 fg_color: str | Tuple[str, str] | None = 'transparent', 
                 border_color: str | Tuple[str, str] | None = None,
                 background_corner_colors: Tuple[str | Tuple[str, str]] | None = None, 
                 overwrite_preferred_drawing_method: str | None = None, **kwargs):
       
        super().__init__(master, 
                        width = width, height = height, 
                        corner_radius = corner_radius, 
                        border_width = border_width, 
                        bg_color = bg_color, 
                        fg_color = fg_color, 
                        border_color = border_color, 
                        background_corner_colors = background_corner_colors, 
                        overwrite_preferred_drawing_method = None, **kwargs)
        self.columnconfigure(0,weight=1)
        self.historyButton=ctk.CTkButton(self,text=text,font=font,text_color=('grey10','grey80'),
                    width=30,fg_color='transparent',anchor='sw',hover_color=('grey60','grey25'),
                    command=lambda: self.Button_command(command[0],text))
        self.historyButton.grid(row=0,column=0,sticky='ew')

        self.XButton = ctk.CTkButton(master=self,text="",fg_color='transparent',hover_color='red',width=20,
                                    image=ctk.CTkImage(dark_image=Image.open("App/Images/light mode/x.png"),
                                                       light_image=Image.open("App/Images/dark mode/x.png")),
                                    command= lambda : self.XCommand(command , text))
        self.XButton.grid(row=0,column=1,sticky='e')

    def Button_command(self,command,text):
        command.delete(0,'end')
        command.insert(0,text)

    def XCommand(self,command, Element):
        command[1](Element)
        command[2]()



App/ToggelFrame.py
import tkinter as tk
import customtkinter as ctk
from typing import Tuple
from PIL import Image,ImageTk


class ToggelMenu(ctk.CTkFrame):
    def __init__(self, master: any, width: int = 200, height: int = 200, 
                 corner_radius: int | str | None = None, 
                 border_width: int | str | None = None, 
                 bg_color: str | Tuple[str, str] = "transparent", 
                 fg_color: str | Tuple[str, str] | None = ('white','#2B2B2B'), 
                 border_color: str | Tuple[str, str] | None = None,
                 background_corner_colors: Tuple[str | Tuple[str, str]] | None = None, 
                 overwrite_preferred_drawing_method: str | None = None, **kwargs):
       
        super().__init__(master, 
                        width = width, height = height, 
                        corner_radius = corner_radius, 
                        border_width = border_width, 
                        bg_color = bg_color, 
                        fg_color = fg_color, 
                        border_color = border_color, 
                        background_corner_colors = background_corner_colors, 
                        overwrite_preferred_drawing_method = None, **kwargs)
        
        self.Font= ctk.CTkFont(family='Bookman Old Style',size=18,weight='bold')

        self.columnconfigure((0,1),weight=1)
        self.rowconfigure(0,weight=1)

        self.Frame= ctk.CTkFrame(master=self,fg_color='transparent')           # First toggel
        self.Frame.grid(row=0,column=0,sticky='nsew')

        self.ButtonFrame=ctk.CTkFrame(self,corner_radius=0)   # second Toggel
        self.ButtonFrame.columnconfigure(0,weight=1)
        
        #============= First Toggel Menu ===============================
        self.Frame.grid_rowconfigure(1, minsize=30)   # empty row with minsize as spacing
        self.Frame.grid_rowconfigure(12, weight=1)  # empty row as spacing
        self.Frame.grid_rowconfigure(14, minsize=15)    # empty row with minsize as spacing
        self.Frame.grid_rowconfigure(16, minsize=30)

        self.Frame.columnconfigure(2,minsize=15)
        
        self.imgmenu = ImageTk.PhotoImage(Image.open('App/Images/light mode/menu.png').resize((40,40)))
        self.imgx = ImageTk.PhotoImage(Image.open('App/Images/light mode/chevron-left.png').resize((50,50)))

        self.Buttonarray=[]  # Button array with [CTkButton class Object , Text , CTkImage class Object]

        #====================== First Toggel Frame ==================== 
        ModeChange = ctk.CTkOptionMenu(master=self.Frame,fg_color=('white','grey25'),text_color=('grey20','grey80'),
                                       button_color=('grey50','grey35'),button_hover_color=('grey40','grey20'),
                                            values=["Light", "Dark", "System"],
                                            command=ctk.set_appearance_mode)
        ModeChange.set('System')
        ModeSwitch = ctk.CTkSwitch(master=self.Frame,text="",width=0,command= lambda :self.change_mode(ModeSwitch))
        ModeSwitch.grid(row=15, column=0,sticky='ew',columnspan=3,padx=20)
   
        label=tk.Label(self.Frame,image=self.imgmenu,text=None,background='#2B2B2B',cursor='hand2')
        label.grid(row=0,column=0,sticky='w',padx=20,pady=20)
        Logo=ctk.CTkLabel(master=self.Frame,text="",image=ctk.CTkImage(light_image=Image.open('App/Images/Logo1.png'),
                            dark_image=Image.open("App/Images/Logo1.png"),size=(250,60)))

        #====================== Second Toggel Frame ====================
        self.ButtonFrame.columnconfigure(0,weight=1)
        self.ButtonFrame.rowconfigure(1,weight=1)

        Xbutton=ctk.CTkButton(self.ButtonFrame,fg_color='transparent',text="",image=self.loadImage(['App/Images/dark mode/x.png','App/Images/light mode/x.png'],30),width=0,command=lambda :Hide_Frame())
        Xbutton.grid(row=0,column=0,sticky='ne')
        #======================Configure toggel ========================

        def open(e):
            label.configure(image=self.imgx)
            Logo.grid(row=0,column=0,sticky='w',padx=5)
            label.grid_configure(column=3,sticky='e')
            if self.Buttonarray:
                for button in self.Buttonarray :
                    button[0].configure(text=button[1])
                    button[0].grid_configure(columnspan=4,sticky='ew')
            ModeSwitch.grid_forget()
            ModeChange.grid(row=15, column=0,padx=20,sticky='ew',columnspan=3)
            label.bind("<Button-1>",close)
            Hide_Frame()

        def close(e=None):
            label.configure(image=self.imgmenu)
            Logo.grid_forget()
            label.grid_configure(column=0,sticky='w')
            if self.Buttonarray:
                for button in self.Buttonarray :
                    button[0].configure(text="")
                    button[0].grid_configure(columnspan=1,sticky='w')
            ModeChange.grid_forget()
            ModeSwitch.grid(row=15, column=0,sticky='e',columnspan=3,padx=20)
            label.bind("<Button-1>",open)

        def Show_Frame():
            self.ButtonFrame.grid(row=0,column=1,sticky='nsew')
            close()
    
        def Hide_Frame():
            self.ButtonFrame.grid_forget() 
        
        label.bind("<Button-1>",open)
        self.Showframe=Show_Frame
    
    def add_button(self,image : list | None,text : str | None ):
        button =ctk.CTkButton(master=self.Frame,text="",image=self.loadImage(image,30),
                              font=self.Font, text_color=("grey5","grey80"),width=30,
                              compound='left',fg_color='transparent',anchor='sw',hover_color=('grey60','grey25'))
        self.Buttonarray.append([button,text,self.loadImage(image,30)])
        self.Update_Button()
        return button

    def Update_Button(self):
        for row,button in enumerate(self.Buttonarray) :
            button[0].grid(row=row+2,column=0,padx=(10,2),pady=10,sticky='w',columnspan=1)

    def loadImage(self,path,size):
        return ctk.CTkImage(light_image=Image.open(path[0]),
                            dark_image=Image.open(path[1]),
                            size=(size,size))  
    
    def change_mode(self,ModeSwitch):
        if ModeSwitch.get() == 0:
            ctk.set_appearance_mode("dark")
        else:
            ctk.set_appearance_mode("light")

    def scale_place(self,coordinate):
        coordinate=(coordinate*1/self._get_widget_scaling())
        return coordinate


if __name__=='__main__':
    def loadImage(path,size):
        return ctk.CTkImage(light_image=Image.open(path),
                            dark_image=Image.open(path),
                            size=(size,size))  
    r=ctk.CTk()
    r.geometry('200x200')
    #ctk.set_appearance_mode('light')
    r.rowconfigure(0, weight=1)
    r.columnconfigure(1, weight=1)
    # ============ create two frames ===
    a=ToggelMenu(master=r)
    a.grid(row=0,column=0,sticky='nsew')
    button=a.add_button(text='Home',image=["App/Images/dark mode/home.png",'App/Images/light mode/x.png'])
    button.configure(command=lambda:a.Showframe())
    
    r.mainloop()

App/mapapp.py
import customtkinter as ctk

import ToggelFrame as TFrame
from HelperFrame import Search_Frame, Layer_Frame, Marker , History
import Map 

ctk.set_default_color_theme("blue")
ctk.set_appearance_mode('dark')

class App(ctk.CTk):

    APP_NAME = "Map's View"
    WIDTH = 800
    HEIGHT = 500

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.title(App.APP_NAME)
        self.geometry(str(App.WIDTH) + "x" + str(App.HEIGHT))
        self.minsize(App.WIDTH, App.HEIGHT)
        self.iconbitmap('App\Images\icon.ico')

        # ============ create two CTkFrames ============

        self.grid_columnconfigure(1, weight=1)
        self.grid_rowconfigure(0, weight=1)

        self.Toggel = TFrame.ToggelMenu(master=self, corner_radius=0, fg_color=None)
        self.Toggel.grid(row=0, column=0, padx=0, pady=0, sticky="nsew")

        self.MapView= Map.Mapwidget(self)
        self.MapView.grid(row=0, column=1, padx=0, pady=0, sticky="nsew")

        #=========================================== Button Functionas =================
        Search=self.Toggel.add_button(text='Search',image=["App/Images/dark mode/search.png","App/Images/light mode/search.png"])
        Search.configure(command=self.Search_Frame)

        Layers=self.Toggel.add_button(text='Tile Server Layout',image=["App/Images/dark mode/layers.png","App/Images/light mode/layers.png"])
        Layers.configure(command=self.layers_Frame)

        Marker=self.Toggel.add_button(text='Markers',image=["App/Images/dark mode/map-pin.png","App/Images/light mode/map-pin.png"])
        Marker.configure(command=self.Configure_marker_frame)

        Offline=self.Toggel.add_button(text='Offline Database',image=["App/Images/dark mode/database.png","App/Images/light mode/database.png"])
        Offline.configure(command=self.offline_Frame)

        #============================================= Functions Frames =====================
        self.SearchFrame=Search_Frame(master=self.Toggel.ButtonFrame,fg_color='transparent')

        #=============== Layers Functions ======================
        Array =["OpenStreetMap","Google normal","Google satellite","Paint Style","Black & White"]
        self.LayerFrame= ctk.CTkFrame(master=self.Toggel.ButtonFrame,fg_color='transparent')
        self.LayerFrame.grid(row=1,column=0,sticky='nsew')
        self.LayerFrame.columnconfigure(0,weight=1)

        for row,text in enumerate(Array):
            layer = Layer_Frame(self.LayerFrame,text=text,font=self.Toggel.Font,command=self.MapView.change_map)
            layer.grid(row=row,column=0,sticky='ew',pady=10,padx=(10,10))
            if row==0: layer.grid_configure(pady=(50,10))
        
        #================ Marker Functions ===================
        self.MarkerFrame= ctk.CTkFrame(master=self.Toggel.ButtonFrame,fg_color='transparent')
        self.MarkerFrame.grid(row=1,column=0,sticky='nsew')

        self.Search_Entry = ctk.CTkEntry(master=self.MarkerFrame,width=250,height=35,
                                            placeholder_text="Type Address")
        self.Search_Entry.grid(row=0, column=0, sticky="we", padx=(12, 12), pady=12,columnspan=2)
        
        self.SetMarkerButton = ctk.CTkButton(master=self.MarkerFrame,text="Set Marker")
        self.SetMarkerButton.grid(pady=20, padx=(20, 20), row=1, column=0)

        self.RemoveMarkerButton = ctk.CTkButton(master=self.MarkerFrame,text="Clear Markers")
        self.RemoveMarkerButton.grid(pady=20, padx=(20, 20), row=1, column=1)

        #================= Offline Frame ==========================
        self.OfflineFrame= ctk.CTkFrame(master=self.Toggel.ButtonFrame,fg_color='transparent')
        self.OfflineFrame.grid(row=1,column=0,sticky='nsew')

        self.OfflineFrame.columnconfigure(0,weight=1)

        frame=ctk.CTkFrame(master=self.OfflineFrame,corner_radius=15)
        frame.grid(row=0,column=0,sticky='ew',padx=(10,10),pady=(20,0))

        frame.columnconfigure(0,weight=1)
        frame.columnconfigure(1,weight=0)

        ctk.CTkLabel(master=frame, text="Offline Maps", font=self.Toggel.Font,anchor='w').grid(row=0,column=0,sticky='ew',padx=(30,20),pady=(10,10))

        offlineSwitch = ctk.CTkSwitch(master=frame,text="",width=20,command= lambda : self.offline_command(offlineSwitch))
        offlineSwitch.grid(row=0,column=1,sticky='ew',padx=(10,20),pady=(10,10))

    def Search_Frame(self):
        Frame= ctk.CTkFrame(master=self.SearchFrame,fg_color='transparent')
        Frame.grid(row=2,column=0,sticky='nsew',columnspan=2)
        Frame.grid_columnconfigure((0,1),weight=1)
        row=2
        for count,history_object in enumerate(self.MapView.search_history) :
            marker=History(Frame,text=history_object,font=self.Toggel.Font,
                          command=[self.SearchFrame.Search_Entry, self.MapView.Remove_search, self.Search_Frame])
            count=count+1
            col= 1 if count%2==0 else 0
            marker.grid(pady=(10, 0), padx=(10,10), row=row,column=col,sticky='ew')
            row= row+1 if count%2==0 else row

        self.SearchFrame.tkraise()
        self.Toggel.Showframe()

    def layers_Frame(self):
        self.LayerFrame.tkraise()
        self.Toggel.Showframe()

    def Configure_marker_frame(self):
        Frame= ctk.CTkFrame(master=self.MarkerFrame,fg_color='transparent')
        Frame.grid(row=2,column=0,sticky='nsew',pady=10,columnspan=2)
        Frame.grid_columnconfigure(0,weight=1)
        for row,marker_object in enumerate(self.MapView.marker_list) :
            marker=Marker(Frame,marker=marker_object,font=self.Toggel.Font,
                          command=[self.MapView.set_position, self.MapView.Remove_marker, self.Configure_marker_frame])
            marker.grid(pady=(10, 0), padx=(20,0), row=row+1,sticky='ew')
        
        self.MarkerFrame.tkraise()
        self.Toggel.Showframe()

    def offline_command(self, offlineMap):

        if offlineMap.get() == 0 :
            self.MapView= Map.Mapwidget(self,use_database_only=False)
        else:
            self.MapView= Map.Mapwidget(self, use_database_only=True, max_zoom=15, database_path="offline_tiles.db")
        
        self.MapView.grid(row=0, column=1, padx=0, pady=0, sticky="nsew")

    def offline_Frame(self):
        self.OfflineFrame.tkraise()
        self.Toggel.Showframe()

    def start(self):
        self.mainloop()
